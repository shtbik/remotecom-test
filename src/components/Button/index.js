import PropTypes from 'prop-types';

import { StyledButton, StyledPrefix } from './styled.index';

export default function Button({ type, variant, children, prefix, ...props }) {
  return (
    <StyledButton type={type} variant={variant} {...props}>
      {prefix && <StyledPrefix>{prefix}</StyledPrefix>}
      {children}
    </StyledButton>
  );
}

Button.defaultProps = {
  type: 'button',
  variant: 'contained',
};

Button.propTypes = {
  /** Native button's tag property */
  type: PropTypes.string,
  /** Various types of view */
  variant: PropTypes.oneOf(['contained', 'outlined']),
  /** Button text content */
  children: PropTypes.node,
  /** The prefix component (Icons, etc.) */
  prefix: PropTypes.node,
};
