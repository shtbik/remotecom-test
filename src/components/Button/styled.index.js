import styled from 'styled-components';

export const StyledButton = styled.button`
  display: inline-flex;
  align-items: center;

  font-size: 1rem;
  font-weight: 500;
  line-height: 1;
  min-height: 44px;
  padding: 10px 25px;
  border: none;
  border-radius: 50px;
  letter-spacing: 0.5px;

  background: var(--colors-irisBlue);
  color: var(--colors-blank);

  &:hover {
    cursor: pointer;
    background: var(--colors-darkBlue);
  }
`;

export const StyledPrefix = styled.span`
  margin-right: 9px;

  path {
    fill: var(--colors-blank);
  }
`;
