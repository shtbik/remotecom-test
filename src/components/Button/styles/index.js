import styled from 'styled-components';

export const ButtonStyled = styled.button`
  display: inline-flex;
  align-items: center;

  font-size: 1rem;
  font-weight: 500;
  line-height: 1;
  min-height: 44px;
  padding: 12px 24px;
  border: none;
  border-radius: 24px;

  background: var(--colors-irisBlue);
  color: var(--colors-blank);
`;
