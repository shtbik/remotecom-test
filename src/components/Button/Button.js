import PropTypes from 'prop-types';

import { ButtonStyled } from './styles';

export default function Button({ children, ...props }) {
  return (
    <ButtonStyled type="button" {...props}>
      {children}
    </ButtonStyled>
  );
}

Button.propTypes = {
  /** Button text content */
  children: PropTypes.node,
};
