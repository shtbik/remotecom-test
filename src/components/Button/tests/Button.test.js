import { cleanup, render, fireEvent } from '@testing-library/react';
import Button from '..';

const text = 'Hello';
const addonText = 'prefix';
const CustomJSX = <div>{addonText}</div>;

describe('Button', () => {
  afterEach(cleanup);

  it('renders correctly', () => {
    const { getByRole, getByText, container } = render(<Button>{text}</Button>);

    const button = getByRole('button');
    expect(button.getAttribute('type')).toBe('button');
    expect(getByText(text)).toBeTruthy();

    expect(container).toMatchSnapshot();
  });

  it('different type', () => {
    const { container } = render(<Button type='submit'>{text}</Button>);

    expect(
      container.querySelector('button[type="submit"]')
    ).toBeInTheDocument();

    expect(container).toMatchSnapshot();
  });

  it('spreads custom attributes', () => {
    const { container, getByRole } = render(
      <Button data-foo='12'>{text}</Button>
    );

    const button = getByRole('button');
    expect(button.getAttribute('data-foo')).toBe('12');

    expect(container).toMatchSnapshot();
  });

  it('firing onClick', () => {
    const clickFn = jest.fn();
    const { container, getByRole } = render(
      <Button onClick={clickFn}>{text}</Button>
    );

    const button = getByRole('button');
    fireEvent.click(button);
    fireEvent.click(button);
    expect(clickFn).toHaveBeenCalledTimes(2);

    expect(container).toMatchSnapshot();
  });

  it('onClick with disabled state', () => {
    const clickFn = jest.fn();
    const { container, getByRole } = render(
      <Button disabled onClick={clickFn}>
        {text}
      </Button>
    );

    const button = getByRole('button');
    fireEvent.click(button);
    fireEvent.click(button);
    expect(clickFn).toHaveBeenCalledTimes(0);

    expect(container).toMatchSnapshot();
  });

  it('add prefix component', () => {
    const { container, getByText } = render(
      <Button prefix={CustomJSX}>{text}</Button>
    );

    expect(container.querySelector('span > div').innerHTML).toBe(addonText);
    expect(getByText(text)).toBeTruthy();

    expect(container).toMatchSnapshot();
  });

  it('loading state', () => {
    const { container, queryByText } = render(<Button loading>{text}</Button>);

    expect(container.querySelector('svg')).toBeInTheDocument();
    expect(queryByText(text)).toBeNull();

    expect(container).toMatchSnapshot();
  });

  it('different variant', () => {
    const { container } = render(<Button variant='outlined'>{text}</Button>);

    expect(container).toMatchSnapshot();
  });
});
