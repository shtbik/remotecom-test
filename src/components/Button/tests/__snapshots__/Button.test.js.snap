// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button add prefix component 1`] = `
<div>
  <button
    class="sc-bdfBwQ Ilmju"
    type="button"
  >
    <span
      class="sc-gsTCUz ldnpWj"
    >
      <div>
        prefix
      </div>
    </span>
    Hello
  </button>
</div>
`;

exports[`Button different type 1`] = `
<div>
  <button
    class="sc-bdfBwQ Ilmju"
    type="submit"
  >
    Hello
  </button>
</div>
`;

exports[`Button different variant 1`] = `
<div>
  <button
    class="sc-bdfBwQ cyOIVg"
    type="button"
  >
    Hello
  </button>
</div>
`;

exports[`Button firing onClick 1`] = `
<div>
  <button
    class="sc-bdfBwQ Ilmju"
    type="button"
  >
    Hello
  </button>
</div>
`;

exports[`Button loading state 1`] = `
<div>
  <button
    class="sc-bdfBwQ Ilmju"
    type="button"
  >
    <svg
      class="sc-dlfnbm bLrFhA"
      variant="contained"
    >
      loading.svg
    </svg>
  </button>
</div>
`;

exports[`Button onClick with disabled state 1`] = `
<div>
  <button
    class="sc-bdfBwQ Ilmju"
    disabled=""
    type="button"
  >
    Hello
  </button>
</div>
`;

exports[`Button renders correctly 1`] = `
<div>
  <button
    class="sc-bdfBwQ Ilmju"
    type="button"
  >
    Hello
  </button>
</div>
`;

exports[`Button spreads custom attributes 1`] = `
<div>
  <button
    class="sc-bdfBwQ Ilmju"
    data-foo="12"
    type="button"
  >
    Hello
  </button>
</div>
`;
